cub/cub/device/dispatch/dispatch_reduce_nondeterministic.cuh

```diff
@@ -0,0 +1,312 @@
+          typename KernelSource = detail::reduce::DeviceReduceNondeterministicKernelSource<
+            typename PolicyHub::MaxPolicy,
+            InputIteratorT,
+            OutputIteratorT,
+            OffsetT,
+            ReductionOpT,
+            InitT,
+            AccumT,
+            TransformOpT>,
+          typename KernelLauncherFactory = detail::TripleChevronFactory>
+struct DispatchReduceNondeterministic
+{
+  static_assert(detail::is_cuda_std_plus_v<ReductionOpT>,
+                "Only plus is currently supported in nondeterministic reduce");
+  //---------------------------------------------------------------------------
+  // Problem state
+  //---------------------------------------------------------------------------
+
+  /// Device-accessible allocation of temporary storage. When `nullptr`, the required allocation
+  /// size is written to `temp_storage_bytes` and no work is done.
+  void* d_temp_storage;
+
+  /// Reference to size in bytes of `d_temp_storage` allocation
+  size_t& temp_storage_bytes;
+
+  /// Pointer to the input sequence of data items
+  InputIteratorT d_in;
+
+  /// Pointer to the output aggregate
+  OutputIteratorT d_out;
+
+  /// Total number of input items (i.e., length of `d_in`)
+  OffsetT num_items;
+
+  /// Binary reduction functor
+  ReductionOpT reduction_op;
+
+  /// The initial value of the reduction
+  InitT init;
+
+  /// CUDA stream to launch kernels within. Default is stream<sub>0</sub>.
+  cudaStream_t stream;
+
+  int ptx_version;
+
+  TransformOpT transform_op;
+
+  KernelSource kernel_source;
+
+  KernelLauncherFactory launcher_factory;
+
+  /**
+   * @brief Invoke a single block block to reduce in-core
+   *
+   * @tparam ActivePolicyT
+   *   Umbrella policy active for the target device
+   *
+   * @tparam AtomicKernelT
+   *   Function type of cub::DeviceReduceAtomicKernel
+   *
+   * @param[in] last_block_kernel
+   *   Kernel function pointer to parameterization of cub::DeviceReduceLastBlockKernel
+   */
+  template <typename ActivePolicyT, typename AtomicKernelT>
+  CUB_RUNTIME_FUNCTION _CCCL_VISIBILITY_HIDDEN _CCCL_FORCEINLINE cudaError_t
+  InvokeAtomicKernel(AtomicKernelT atomic_kernel, ActivePolicyT active_policy = {})
+  {
+    // This memory is not actually needed but we keep it to make sure the API is consistent
+    if (d_temp_storage == nullptr)
+    {
+      temp_storage_bytes = 1;
+      return cudaSuccess;
+    }
+
+    // Init regular kernel configuration
+    detail::KernelConfig reduce_config;
+    cudaError_t error =
+      CubDebug(reduce_config.Init(atomic_kernel, active_policy.ReduceNondeterministic(), launcher_factory));
+    if (cudaSuccess != error)
+    {
+      return error;
+    }
+
+    const int reduce_grid_size = static_cast<int>(::cuda::ceil_div(num_items, reduce_config.tile_size));
+
+    error = CubDebug(launcher_factory.MemsetAsync(d_out, 0, 1, kernel_source.InitSize(), stream));
```

Comment:
Q: What happens if `InitSize()` is not a power of two or smaller than 4 and the driver API is used? For example, when reducing a `ushort3` or a `short`.

Reply:
Great!